### PR TITLE
Fix Memory Leak with Weak References

### PR DESCRIPTION
**Pull request by wpm**
PDF parsing leaks memory. If you surround a parsing operation with gc.set_debug(gc.DEBUG_SAVEALL) and gc.garbage you would see a large number of PdfMiner objects that are unable to be garbage collected.

The PSBaseParser keeps track of its parse state by having a _parse1 instance variable assigned to member functions of the same instance. This leads to circular references which causes the memory leaks.

The fix is to change all the _parse1 assignments to be weakref.WeakMethod objects.

This is an alternate implementation for pull request #687 . I like this one better because it involves less code change. 

**Pull request**

Thanks for improving pdfminer.six! Please include the following information to
help us discuss and merge this PR:

- Fixes a memory leak in PDF parsing
- Make function assignments to weakrefs
- #580 

**Checklist**

- [ ] I have added tests that prove my fix is effective or that my feature 
  works
- [ ] I have added docstrings to newly created methods and classes
- [ ] I have optimized the code at least one time after creating the initial 
  version
- [x] I have updated the [README.md](../README.md) or I am verified that this
  is not necessary
- [x] I have updated the [readthedocs](../docs/source) documentation or I 
  verified that this is not necessary
- [x] I have added a concise human-readable description of the change to 
  [CHANGELOG.md](../CHANGELOG.md)

### CODE REVIEW

File: pdfminer/psparser.py [hunk 1]
```diff
@@ -217,7 +218,7 @@ def seek(self, pos: int) -> None:
         self.buf = b''
         self.charpos = 0
         # reset the status for nexttoken()
-        self._parse1 = self._parse_main
+        self._parse1 = weakref.WeakMethod(self._parse_main)
```
Comment (0xabu) [inline pdfminer/psparser.py:221]: I'd be tempted to add a brief comment here explaining the use of a weak ref to avoid a cyclic reference.

**Review by 0xabu — DISMISSED** [review]

**Comment by pietermarsman** [comment]
I'm not sure if this PR's fixes the issue. 

I use this code for testing (inspired on yours):

```python
from pdfminer.high_level import extract_text
from memory_profiler import profile
import gc


@profile
def main():
    extract_text("samples/simple1.pdf")
    gc.collect()

main()
```

The output on develop is:

```
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     6     24.1 MiB     24.1 MiB           1   @profile
     7                                         def main():
     8     24.7 MiB      0.6 MiB           1       extract_text("samples/simple1.pdf")
     9     24.7 MiB      0.0 MiB           1       gc.collect()
```


The output on this branch is:

```
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     6     23.4 MiB     23.4 MiB           1   @profile
     7                                         def main():
     8     23.9 MiB      0.5 MiB           1       extract_text("samples/simple1.pdf")
     9     23.9 MiB      0.0 MiB           1       gc.collect()
```

To be fair, I'm not entirely sure how to read the output. But I think we expect the increment at `gc.collect()` to be of the same size as the increment at `extract_text`, only negative. But it does not show any difference.

**Comment by pietermarsman** [comment]
The same is true for #687. 

Let me know if I'm interpreting the output wrongly.

**Comment by pietermarsman** [comment]
FYI, I've changed this MR to merge into master. The develop branch will be removed, because soon we will work with version tags to indicate the releases and the distinction between develop and master becomes obsolete.

**Comment by pietermarsman** [comment]
Closing due to inactivity

**Pull request closed without merging**